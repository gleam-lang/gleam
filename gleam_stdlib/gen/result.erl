-module(result).
-compile(no_auto_import).
-include_lib("eunit/include/eunit.hrl").

-export([is_ok/1, is_error/1, map/2, map_error/2, flatten/1, flat_map/2, unwrap/2]).

is_ok(Result) ->
    case Result of
        {error, _} ->
            false;

        {ok, _} ->
            true
    end.

-ifdef(TEST).
is_ok_test() ->
    expect:true(is_ok({ok, 1})),
    expect:false(is_ok({error, 1})).
-endif.

is_error(Result) ->
    case Result of
        {ok, _} ->
            false;

        {error, _} ->
            true
    end.

-ifdef(TEST).
is_error_test() ->
    expect:false(is_error({ok, 1})),
    expect:true(is_error({error, 1})).
-endif.

map(Result, Fun) ->
    case Result of
        {ok, X} ->
            {ok, Fun(X)};

        {error, _} ->
            Result
    end.

-ifdef(TEST).
map_test() ->
    (fun(Capture1) ->
        expect:equal(Capture1, {ok, 2})
    end)((fun(Capture1) -> map(Capture1, fun(X) -> X + 1 end) end)({ok, 1})),
    (fun(Capture1) ->
        expect:equal(Capture1, {error, 1})
    end)((fun(Capture1) -> map(Capture1, fun(X) -> X + 1 end) end)({error, 1})).
-endif.

map_error(Result, Fun) ->
    case Result of
        {ok, _} ->
            Result;

        {error, Error} ->
            {error, Fun(Error)}
    end.

-ifdef(TEST).
map_error_test() ->
    (fun(Capture1) ->
        expect:equal(Capture1, {ok, 1})
    end)((fun(Capture1) ->
             map_error(Capture1, fun(X) -> X + 1 end)
         end)({ok, 1})),
    (fun(Capture1) ->
        expect:equal(Capture1, {error, 2})
    end)((fun(Capture1) ->
             map_error(Capture1, fun(X) -> X + 1 end)
         end)({error, 1})).
-endif.

flatten(Result) ->
    case Result of
        {ok, X} ->
            X;

        {error, Error} ->
            {error, Error}
    end.

-ifdef(TEST).
flatten_test() ->
    (fun(Capture1) ->
        expect:equal(Capture1, {ok, 1})
    end)(flatten({ok, {ok, 1}})),
    (fun(Capture1) ->
        expect:equal(Capture1, {error, 1})
    end)(flatten({ok, {error, 1}})),
    (fun(Capture1) ->
        expect:equal(Capture1, {error, 1})
    end)(flatten({error, 1})),
    (fun(Capture1) ->
        expect:equal(Capture1, {error, {error, 1}})
    end)(flatten({error, {error, 1}})).
-endif.

flat_map(Result, Fun) ->
    case Result of
        {ok, X} ->
            case Fun(X) of
                {ok, Y} ->
                    {ok, Y};

                {error, Y1} ->
                    {error, Y1}
            end;

        {error, _} ->
            Result
    end.

-ifdef(TEST).
flat_map_test() ->
    (fun(Capture1) ->
        expect:equal(Capture1, {error, 1})
    end)((fun(Capture1) ->
             flat_map(Capture1, fun(X) -> {ok, X + 1} end)
         end)({error, 1})),
    (fun(Capture1) ->
        expect:equal(Capture1, {ok, 2})
    end)((fun(Capture1) ->
             flat_map(Capture1, fun(X) -> {ok, X + 1} end)
         end)({ok, 1})),
    (fun(Capture1) ->
        expect:equal(Capture1, {error, 1})
    end)((fun(Capture1) ->
             flat_map(Capture1, fun(Unused) -> {error, 1} end)
         end)({ok, 1})).
-endif.

unwrap(Result, Default) ->
    case Result of
        {ok, V} ->
            V;

        {error, _} ->
            Default
    end.

-ifdef(TEST).
unwrap_test() ->
    (fun(Capture1) -> expect:equal(Capture1, 1) end)(unwrap({ok, 1}, 50)),
    (fun(Capture1) ->
        expect:equal(Capture1, 50)
    end)(unwrap({error, <<"nope">>}, 50)).
-endif.
