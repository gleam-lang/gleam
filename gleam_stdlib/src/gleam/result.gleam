// Result represents the result of something that may succeed or fail.
// `Ok` means it was successful, `Error` means it failed.

pub fn is_ok(result) {
  case result {
  | Error(_) -> False
  | Ok(_) -> True
  }
}

pub fn is_error(result) {
  case result {
  | Ok(_) -> False
  | Error(_) -> True
  }
}

pub fn map(result, with fun) {
  case result {
  | Ok(x) -> Ok(fun(x))
  | Error(e) -> Error(e)
  }
}

pub fn map_error(result, with fun) {
  case result {
  | Ok(x) -> Ok(x)
  | Error(error) -> Error(fun(error))
  }
}

pub fn flatten(result) {
  case result {
  | Ok(x) -> x
  | Error(error) -> Error(error)
  }
}

pub fn then(result, apply fun) {
  case result {
  | Ok(x) -> fun(x)
  | Error(e) -> Error(e)
  }
}

pub fn unwrap(result, or default) {
  case result {
  | Ok(v) -> v
  | Error(_) -> default
  }
}
