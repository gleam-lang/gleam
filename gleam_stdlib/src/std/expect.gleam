// TODO: Move this module into another package so it can be used as a
// dep only in test.

pub external type Expectation;

pub external fn equal(a, a) -> Expectation = "gleam__stdlib" "expect_equal";

pub external fn not_equal(a, a) -> Expectation = "gleam__stdlib" "expect_not_equal";

pub external fn true(Bool) -> Expectation = "gleam__stdlib" "expect_true";

pub external fn false(Bool) -> Expectation = "gleam__stdlib" "expect_false";

pub external fn is_ok(Result(a, b)) -> Expectation = "gleam__stdlib" "expect_is_ok";

pub external fn is_error(Result(a, b)) -> Expectation = "gleam__stdlib" "expect_is_error";

pub fn fail() {
  true(False)
}
