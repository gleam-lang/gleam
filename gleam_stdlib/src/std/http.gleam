// HTTP standard method as defined by RFC 2616, and PATCH which is defined by
// RFC 5789.
//
pub enum Method =
  | Get
  | Post
  | Head
  | Put
  | Delete
  | Trace
  | Connect
  | Options
  | Patch
