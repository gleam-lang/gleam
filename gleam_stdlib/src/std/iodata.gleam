pub external type Iodata;

pub external fn prepend(Iodata, String) -> Iodata =
  "gleam__stdlib" "iodata_prepend";

pub external fn append(Iodata, String) -> Iodata =
  "gleam__stdlib" "iodata_append";

pub external fn prepend_iodata(Iodata, Iodata) -> Iodata =
  "gleam__stdlib" "iodata_prepend";

pub external fn append_iodata(Iodata, Iodata) -> Iodata =
  "gleam__stdlib" "iodata_append";

pub external fn from_strings(List(String)) -> Iodata =
  "gleam__stdlib" "identity";

pub external fn concat(List(Iodata)) -> Iodata =
  "gleam__stdlib" "identity";

pub external fn new(String) -> Iodata =
  "gleam__stdlib" "identity";

pub external fn to_string(Iodata) -> String =
  "erlang" "iolist_to_binary";

pub external fn byte_size(Iodata) -> Int =
  "erlang" "iolist_size";

pub external fn from_float(Float) -> Iodata =
  "io_lib_format" "fwrite_g";

pub external fn lowercase(Iodata) -> Iodata = "string" "lowercase"

pub external fn uppercase(Iodata) -> Iodata = "string" "uppercase"

pub external fn reverse(Iodata) -> Iodata = "string" "reverse"

enum Direction =
  | All

external fn erl_split(Iodata, String, Direction) -> List(Iodata) =
  "string" "split"

pub fn split(iodata, on) {
  erl_split(iodata, on, All)
}

external fn erl_replace(Iodata, String, String, Direction) -> Iodata =
  "string" "replace"

pub fn replace(iodata, pattern, replacement) {
  erl_replace(iodata, pattern, replacement, All)
}

pub external fn is_equal(Iodata, Iodata) -> Bool = "string" "equal"

pub external fn is_empty(Iodata) -> Bool = "string" "is_empty"
