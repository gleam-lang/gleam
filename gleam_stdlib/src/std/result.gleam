// Result represents the result of something that may succeed or fail.
// `Ok` means it was successful, `Error` means it failed.

pub fn is_ok(result) {
  case result {
  | Error(_) -> False
  | Ok(_) -> True
  }
}

pub fn is_error(result) {
  case result {
  | Ok(_) -> False
  | Error(_) -> True
  }
}

pub fn map(result, fun) {
  case result {
  | Ok(x) -> Ok(fun(x))
  | Error(e) -> Error(e)
  }
}

pub fn map_error(result, fun) {
  case result {
  | Ok(_) -> result
  | Error(error) -> Error(fun(error))
  }
}

pub fn flatten(result) {
  case result {
  | Ok(x) -> x
  | Error(error) -> Error(error)
  }
}

pub fn then(result, fun) {
  case result {
  | Ok(x) -> fun(x)
  | Error(e) -> Error(e)
  }
}

pub fn unwrap(result, default) {
  case result {
  | Ok(v) -> v
  | Error(_) -> default
  }
}
