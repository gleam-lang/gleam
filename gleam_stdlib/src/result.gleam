import expect

// Result represents the result of something that may succeed or fail.
// `Ok` means it was successful, `Error` means it failed.
//
pub enum Result(error, value) =
  | Ok(value)
  | Error(error)
;

pub fn is_ok(result) {
  case result {
  | Error(_) -> False
  | Ok(_) -> True
  }
}

test is_ok {
  is_ok(Ok(1)) |> expect:true
  is_ok(Error(1)) |> expect:false
}

pub fn is_error(result) {
  case result {
  | Ok(_) -> False
  | Error(_) -> True
  }
}

test is_error {
  is_error(Ok(1))
    |> expect:false

  is_error(Error(1))
    |> expect:true
}

pub fn map(result, fun) {
  case result {
  | Ok(x) -> Ok(fun(x))
  | Error(_) -> result
  }
}

test map {
  Ok(1)
    |> map(_, fn(x) { x + 1 })
    |> expect:equal(_, Ok(2))

  Error(1)
    |> map(_, fn(x) { x + 1 })
    |> expect:equal(_, Error(1))
}

pub fn map_error(result, fun) {
  case result {
  | Ok(_) -> result
  | Error(error) -> Error(fun(error))
  }
}

test map_error {
  Ok(1)
    |> map_error(_, fn(x) { x + 1 })
    |> expect:equal(_, Ok(1))

  Error(1)
    |> map_error(_, fn(x) { x + 1 })
    |> expect:equal(_, Error(2))
}

pub fn flatten(result) {
  case result {
  | Ok(x) -> x
  | Error(error) -> Error(error)
  }
}

test flatten {
  flatten(Ok(Ok(1)))
    |> expect:equal(_, Ok(1))

  flatten(Ok(Error(1)))
    |> expect:equal(_, Error(1))

  flatten(Error(1))
    |> expect:equal(_, Error(1))

  flatten(Error(Error(1)))
    |> expect:equal(_, Error(Error(1)))
}

pub fn flat_map(result, fun) {
  case result {
  | Ok(x) ->
      case fun(x) {
      | Ok(y) -> Ok(y)
      | Error(y) -> Error(y)
      }
  | Error(_) -> result
  }
}

test flat_map {
  Error(1)
    |> flat_map(_, fn(x) { Ok(x + 1) })
    |> expect:equal(_, Error(1))

  Ok(1)
    |> flat_map(_, fn(x) { Ok(x + 1) })
    |> expect:equal(_, Ok(2))

  Ok(1)
    |> flat_map(_, fn(unused) { Error(1) })
    |> expect:equal(_, Error(1))
}

pub fn unwrap(result, default) {
  case result {
  | Ok(v) -> v
  | Error(_) -> default
  }
}

test unwrap {
  unwrap(Ok(1), 50)
    |> expect:equal(_, 1)

  unwrap(Error("nope"), 50)
    |> expect:equal(_, 50)
}
