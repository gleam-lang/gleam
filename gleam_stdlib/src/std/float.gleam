import std/iodata

pub enum NotAFloat =
  | NotAFloat

pub external fn parse(String) -> Result(Float, NotAFloat) =
  "gleam__stdlib" "parse_float";

pub fn to_string(f) {
  f
  |> iodata:from_float
  |> iodata:to_string
}

pub external fn ceiling(Float) -> Float = "math" "ceil";

pub external fn floor(Float) -> Float = "math" "floor";

pub external fn round(Float) -> Int = "erlang" "round";

pub external fn truncate(Float) -> Int = "erlang" "trunc";
