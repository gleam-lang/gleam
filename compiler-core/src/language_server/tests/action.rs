use std::assert_eq;

use crate::{language_server::engine, line_numbers::LineNumbers};
use lsp_types::{
    CodeActionContext, CodeActionParams, PartialResultParams, Position, Range,
    TextDocumentIdentifier, Url, WorkDoneProgressParams, WorkspaceEdit,
};

use super::*;

const TEST_FILE_PATH: &str = match cfg!(target_family = "windows") {
    true => r"\\?\C:\src\app.gleam",
    false => "/src/app.gleam",
};

fn test_file_url() -> Url {
    Url::from_file_path(Utf8PathBuf::from(TEST_FILE_PATH)).expect("file path is valid url")
}

fn engine_response(src: &str, line: u32) -> engine::Response<Option<Vec<lsp_types::CodeAction>>> {
    let io = LanguageServerTestIO::new();
    let mut engine = setup_engine(&io);

    // inject stdlib stubs
    _ = io.src_module("list", "");
    _ = io.src_module(
        "result",
        "pub fn is_ok() {}\npub fn is_err() {}\npub fn all() {}",
    );
    _ = io.src_module("map", "pub type Map(key, value)\npub fn delete() {}");
    _ = io.src_module("option", "");

    _ = io.src_module("app", src);
    engine.compile_please().result.expect("compiled");

    let params = CodeActionParams {
        text_document: TextDocumentIdentifier::new(test_file_url()),
        context: CodeActionContext {
            diagnostics: vec![],
            only: None,
            trigger_kind: None,
        },
        range: Range::new(Position::new(0, 0), Position::new(line + 1, 0)),
        work_done_progress_params: WorkDoneProgressParams {
            work_done_token: None,
        },
        partial_result_params: PartialResultParams {
            partial_result_token: None,
        },
    };

    engine.action(params)
}

const REMOVE_UNUSED_IMPORTS_TITLE: &str = "Remove unused imports";
const REMOVE_REDUNDANT_TUPLES: &str = "Remove redundant tuples";

fn apply_first_code_action_with_title(src: &str, line: u32, title: &str) -> String {
    let response = engine_response(src, line)
        .result
        .unwrap()
        .and_then(|actions| actions.into_iter().find(|action| action.title == title));
    if let Some(action) = response {
        apply_code_action(src, &test_file_url(), &action)
    } else {
        panic!("No code action produced by the engine")
    }
}

fn apply_code_action(src: &str, url: &Url, action: &lsp_types::CodeAction) -> String {
    match &action.edit {
        Some(WorkspaceEdit { changes, .. }) => match changes {
            Some(changes) => apply_code_edit(src, url, changes),
            None => panic!("No text edit found"),
        },
        _ => panic!("No workspace edit found"),
    }
}

// This function replicates how the text editor applies TextEdit
fn apply_code_edit(
    src: &str,
    url: &Url,
    changes: &HashMap<Url, Vec<lsp_types::TextEdit>>,
) -> String {
    let mut result = src.to_string();
    let line_numbers = LineNumbers::new(src);
    let mut offset = 0;
    for (change_url, change) in changes {
        if url != change_url {
            panic!("Unknown url {}", change_url)
        }
        for edit in change {
            let start =
                line_numbers.byte_index(edit.range.start.line, edit.range.start.character) - offset;
            let end =
                line_numbers.byte_index(edit.range.end.line, edit.range.end.character) - offset;
            let range = (start as usize)..(end as usize);
            offset += end - start - edit.new_text.len() as u32;
            result.replace_range(range, &edit.new_text);
        }
    }
    result
}

#[test]
fn test_remove_unused_simple() {
    let code = "
// test
import // comment
  list as lispy
import result
import option

pub fn main() {
  result.is_ok
}
";
    let expected = "
// test
import result

pub fn main() {
  result.is_ok
}
";
    assert_eq!(
        apply_first_code_action_with_title(code, 2, REMOVE_UNUSED_IMPORTS_TITLE),
        expected.to_string()
    )
}

#[test]
fn test_remove_unused_start_of_file() {
    let code = "import option
import result

pub fn main() {
  result.is_ok
}
";
    let expected = "import result

pub fn main() {
  result.is_ok
}
";
    assert_eq!(
        apply_first_code_action_with_title(code, 2, REMOVE_UNUSED_IMPORTS_TITLE),
        expected.to_string()
    )
}

#[test]
fn test_remove_unused_alias() {
    let code = "
// test
import result.{is_ok} as res
import option

pub fn main() {
  is_ok
}
";
    let expected = "
// test
import result.{is_ok}%SPACE%

pub fn main() {
  is_ok
}
";
    assert_eq!(
        apply_first_code_action_with_title(code, 2, REMOVE_UNUSED_IMPORTS_TITLE),
        expected.replace("%SPACE%", " ")
    )
}

#[test]
fn test_remove_redundant_tuple_in_case_subject_simple() {
    let code = "
pub fn main() {
  case #(1) { #(a) -> 0 }
  case #(1, 2) { #(a, b) -> 0 }
}
";

    let expected = "
pub fn main() {
  case 1 { a -> 0 }
  case 1, 2 { a, b -> 0 }
}
";

    assert_eq!(
        apply_first_code_action_with_title(code, 7, REMOVE_REDUNDANT_TUPLES),
        expected
    );
}

#[test]
fn test_remove_redundant_tuple_in_case_subject_nested() {
    let code = "
pub fn main() {
  case #(case #(0) { #(a) -> 0 }) { #(b) -> 0 }
}
";

    let expected = "
pub fn main() {
  case case 0 { a -> 0 } { b -> 0 }
}
";

    assert_eq!(
        apply_first_code_action_with_title(code, 7, REMOVE_REDUNDANT_TUPLES),
        expected
    );
}

#[test]
fn test_remove_redundant_tuple_in_case_retain_extras() {
    let code = "
pub fn main() {
  case
    #(
      // first comment
      1,
      // second comment
      2,
      3 // third comment before comma

      ,

      // fourth comment after comma

    )
  {
    #(
      // first comment
      a,
      // second comment
      b,
      c // third comment before comma

      ,

      // fourth comment after comma

    ) -> 0
  }
}
";

    let result = apply_first_code_action_with_title(code, 20, REMOVE_REDUNDANT_TUPLES);

    insta::assert_snapshot!(result);
}

#[test]
fn test_remove_redundant_tuple_in_case_subject_ignore_empty_tuple() {
    let code = "
pub fn main() {
  case #() { #() -> 0 }
}
";

    assert!(engine_response(code, 11)
        .result
        .expect("ok response")
        .is_none());
}

#[test]
fn test_remove_redundant_tuple_in_case_subject_only_safe_remove() {
    let code = "
pub fn main() {
  case #(0), #(1) {
    #(1), #(b) -> 0
    a, #(0) -> 1 // The first of this clause is not a tuple
    #(a), #(b) -> 2
  }
}
";

    let expected = "
pub fn main() {
  case #(0), 1 {
    #(1), b -> 0
    a, 0 -> 1 // The first of this clause is not a tuple
    #(a), b -> 2
  }
}
";

    assert_eq!(
        apply_first_code_action_with_title(code, 11, REMOVE_REDUNDANT_TUPLES),
        expected
    );
}

/* TODO: implement qualified unused location
#[test]
fn test_remove_unused_qualified_action() {
    let code = "
// test
import map.{Map, delete}
";
    let expected = "
// test

";
    assert_eq!(remove_unused_action(code), expected.to_string())
}

#[test]
fn test_remove_unused_qualified_partial_action() {
    let code = "
// test
import result.{is_ok, is_err}

pub fn main() {
  is_ok
}
";
    let expected = "
// test
import result.{is_ok}

pub fn main() {
  is_ok
}
";
    assert_eq!(remove_unused_action(code), expected.to_string())
}

#[test]
fn test_remove_unused_qualified_partial2_action() {
    let code = "
// test
import result.{all, is_ok, is_err}

pub fn main() {
  is_ok
}
";
    let expected = "
// test
import result.{ is_ok}

pub fn main() {
  is_ok
}
";
    assert_eq!(remove_unused_action(code), expected.to_string())
}

#[test]
fn test_remove_unused_qualified_partial3_action() {
    let code = "
// test
import result.{all, is_ok, is_err} as res

pub fn main() {
  is_ok
}
";
    let expected = "
// test
import result.{ is_ok} as res

pub fn main() {
  is_ok
}
";
    assert_eq!(remove_unused_action(code), expected.to_string())
}
*/
