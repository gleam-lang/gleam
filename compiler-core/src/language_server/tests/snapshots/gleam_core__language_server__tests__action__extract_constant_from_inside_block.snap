---
source: compiler-core/src/language_server/tests/action.rs
expression: "pub fn main() {\n  let fahrenheit = {\n    let degrees = 64 + 32\n    degrees\n  }\n}"
---
----- BEFORE ACTION
pub fn main() {
  let fahrenheit = {
    let degrees = 64 + 32
                       â†‘ 
    degrees
  }
}


----- AFTER ACTION
const int = 32

pub fn main() {
  let fahrenheit = {
    let degrees = 64 + int
    degrees
  }
}
